### PR TITLE
Enable stricter lints

### LINKED ISSUE #1132: Enable strict lints and checks
1. We must enable as maximum "compile time" checks as possible to catch potential problems early (not on runtime).
2. We also must forbid merging when there are failed links or formatting issues.
3. Ideally we want to have a pre-push hook that checks the lints, so that the author of the PR knows about issues in advance (before pushing the branch or creating a PR). This would avoid not particularly useful "fix formatting", "fix lints" and other commits.

### PR DESCRIPTION
**Pull request by daniel-abramov**
This is an attempt to fix https://github.com/vector-im/element-call/issues/1132

### Rationale

The main idea is that after this PR we'll have a decent set of TypeScript strictness and we'll forbid the `@ts-ignore`, so that we know that our code is sound [type-wise].

### Important Notes

1. I haven't enabled the strictest lints possible for the reasons discussed in a team meeting (i.e. too many errors and incompatibilities).
2. I deliberately used `@ts-ignore` in certain parts where the fix of the issue looked untrivial or where I'm not familiar with fixing it (feel free to submit a follow-up or push a change to the branch if you know a quick and easy way to fix it). This is primarily seen in `VideoGrid`, `NewVideoGrid`, `rageshake.ts`, `GroupCallInspector`. I also did not want to touch the video grid logic much since @robintown works on a video grid.
3. In many cases I had to use `!` to assert the type in cases where it looked like we were expecting the actual type.
4. I have refactored the `ClientContext` and tried to keep the compatibility layer for the components that expected a different "plain" representation of a client state.
5. I [intentionally] did not fix some "logical errors" (weirdness and unnecessary complications of code, i.e. accidental complexity) in some parts as that would result in a massive PR with possibly more regressions, this PR is the first step to enable clients and make sure that we can start building on top of a stricter type system.
6. ~~I have not yet performed any smoke tests, so things may fail, especially if some assumptions (e.g. those mentioned in (3)) were wrong.~~

### CODE REVIEW

File: src/Avatar.tsx
Comment (robintown) [inline src/Avatar.tsx]: `resolveAvatarSrc` also checks for whether the client is undefined, so you can remove that redundant check above now

File: src/UserMenuContainer.tsx
Comment (robintown) [inline src/UserMenuContainer.tsx]: For future reference you can write this more concisely as `logout?.()`

File: src/home/useGroupCallRooms.ts
Comment (robintown) [inline src/home/useGroupCallRooms.ts:102]: Oh nice, I never guessed that we actually implemented showing the participants like this, but then broke the feature by messing up the types

File: src/profile/useProfile.ts
Comment (robintown) [inline src/profile/useProfile.ts]:
```suggestion
export function useProfile(client: MatrixClient | undefined) {
```

Let's do this so the caller can't completely forget to pass something for the `client` argument

File: src/room/GridLayoutMenu.tsx
Comment (robintown) [inline src/room/GridLayoutMenu.tsx]: Could we follow the established pattern here of wrapping all callbacks in `useCallback` as a general rule?

File: src/room/GroupCallLoader.tsx
Comment (robintown) [inline src/room/GroupCallLoader.tsx]: I believe it's safe to use an assertion `groupCall!` here since both `loading` and `error` are checked above

File: src/room/GroupCallView.tsx
Comment (robintown) [inline src/room/GroupCallView.tsx]:
```suggestion
          widget!.api.setAlwaysOnScreen(true),
          widget!.api.transport.reply(ev.detail, {}),
```

due to the `if (widget)` above

File: src/room/GroupCallView.tsx
Comment (robintown) [inline src/room/GroupCallView.tsx]:
```suggestion
        widget!.lazyActions.off(ElementWidgetActions.JoinCall, onJoin);
```

File: src/room/GroupCallView.tsx
Comment (robintown) [inline src/room/GroupCallView.tsx]:
```suggestion
        await widget!.api.transport.reply(ev.detail, {});
        widget!.api.setAlwaysOnScreen(false);
```

File: src/room/GroupCallView.tsx
Comment (robintown) [inline src/room/GroupCallView.tsx]:
```suggestion
        widget!.lazyActions.off(ElementWidgetActions.HangupCall, onHangup);
```

File: src/room/InCallView.tsx
Comment (robintown) [inline src/room/InCallView.tsx]:
```suggestion
        await widget!.api.transport.reply(ev.detail, {});
```

File: src/room/InCallView.tsx
Comment (robintown) [inline src/room/InCallView.tsx]:
```suggestion
        await widget!.api.transport.reply(ev.detail, {});
```

File: src/room/InCallView.tsx
Comment (robintown) [inline src/room/InCallView.tsx]:
```suggestion
        widget!.lazyActions.off(ElementWidgetActions.TileLayout, onTileLayout);
        widget!.lazyActions.off(
```

File: src/home/CallTypeDropdown.tsx
Comment (robintown) [inline src/home/CallTypeDropdown.tsx]: this could also use `useCallback`

File: src/button/LinkButton.tsx
Comment (robintown) [inline src/button/LinkButton.tsx]: As long as we're using `sizeToClassName` to prepare for the possibility of there being more than one button size, we should probably not hardcode `lg` here

File: src/UserMenuContainer.tsx
Comment (robintown) [inline src/UserMenuContainer.tsx]: Hiding the menu if the user doesn't have an avatar set seems wrong

File: src/UserMenu.tsx
Comment (robintown) [inline src/UserMenu.tsx]: It's probably fine to leave this for a future PR to avoid making this PR too big, but we should consider changing `PopoverMenuTrigger` to take this function as a normal prop rather than as a child, so it's easier to write types for

File: src/ClientContext.tsx
Comment (robintown) [inline src/ClientContext.tsx]: This combination of `useState` + `useEffect` could be replaced with a single `useMemo` hook to achieve the same thing with fewer renders

**Review by robintown — COMMENTED** [review]
Thanks a lot, this will improve the developer experience of working on Element Call significantly! I'm not super happy about all the `@ts-ignore` comments we have now, but I guess working iteratively to improve our types is a good strategy.

**Comment by daniel-abramov** [comment]
> Thanks a lot, this will improve the developer experience of working on Element Call significantly! I'm not super happy about all the `@ts-ignore` comments we have now, but I guess working iteratively to improve our types is a good strategy.

**Fully agree.** AFAIK one of the strategies that some teams use to approach a component that has too many 'violations' of the strictness, is to suppress all of the warnings/errors with a `@ts-ignore` at the first stage and introduce a rule "no new `@ts-ignore` anymore", so that at least the new code benefits from a stricter set of rules, while at the same time consider removing the existing `@ts-ignore` when working on features/further improvements.

I decided to go a step further and fix those that looked easy to set, venturing into `ClientContext` refactoring for a better way to fix them. Still, it looked like it takes too much time, so I thought it would be better to suppress some of them that I can't fix quickly or don't have enough context to fix.

File: src/Avatar.tsx
Comment (daniel-abramov) [reply src/Avatar.tsx]: Ah, but the type on the `resovleAvatarSrc` was not optional (though it used optional checks inside the function 🙂), so the type error was triggered and I did not check how the function looks. Now, after looking, I realized that the function could be inlined since everything is checked. Thanks for pointing this out! Fixed.

File: src/UserMenuContainer.tsx
Comment (daniel-abramov) [reply src/UserMenuContainer.tsx]: Thank you! My bad, was probably too lazy to check how it must be written. Fixed.

File: src/profile/useProfile.ts
Comment (daniel-abramov) [reply src/profile/useProfile.ts]: That makes sense, agree! Fixed.

File: src/room/GridLayoutMenu.tsx
Comment (daniel-abramov) [reply src/room/GridLayoutMenu.tsx]: Good catch, fixed!

OOC: Do you by chance know if there is a lint for that or something similar so that such things could be caught in advance? I'm always forgetting about these sorts of things (not a frontend/React developer 🙂).

File: src/home/CallTypeDropdown.tsx
Comment (daniel-abramov) [reply src/home/CallTypeDropdown.tsx]: Agreed, done!

File: src/button/LinkButton.tsx
Comment (daniel-abramov) [reply src/button/LinkButton.tsx]: Agree, fixed.

File: src/UserMenuContainer.tsx
Comment (daniel-abramov) [reply src/UserMenuContainer.tsx]: Good point, fixed!

File: src/UserMenu.tsx
Comment (daniel-abramov) [reply src/UserMenu.tsx]: I'm not an expert, but sounds sensible 🙂 I'm definitely not a fan of `any` and weaker types.

File: src/ClientContext.tsx
Comment (daniel-abramov) [reply src/ClientContext.tsx]: My poor React skills, fixed 🙂 It looks much better with `useMemo()`, thanks!

**Comment by daniel-abramov** [comment]
**UPD.** Rebased against `livekit` to resolve conflicts.

File: src/room/GroupCallLoader.tsx
Comment (daniel-abramov) [reply src/room/GroupCallLoader.tsx]: Ah, so the fields in the structure were "either/or`? - Ok, now I see, checked the source code of the hook. Thanks for pointing this out!

After checking the internals of a function, I could not stand an implicit (not properly modeled) state, so I decided to refactor the hook to make it more explicit and prevent impossible states. The whole thing [looks much more elegant now IMHO](https://github.com/vector-im/element-call/pull/1173/commits/03ba4ac45ae061c672fab7062f99efa9d7aeedae#diff-25657a65748b36f55c850afaa4a56aab67db55013fe52e3eb3ecda1a1fe2cb56R48) and also strongly typed. Hope you like it too 🙂

File: src/room/GroupCallView.tsx
Comment (daniel-abramov) [reply src/room/GroupCallView.tsx]: Just to double-check: we guarantee that the `widget` cannot be modified otherwise, right?

The reason I'm asking is that at this place the `widget` is used from the callback that is defined inside the `if`, not from the `if` block by itself, hence the warning from a transpiler, i.e. by the point the callback is called, the widget may become `null` again. The `on()` action that subscribes to the event does not trigger this warning since it's called directly inside an `if` (i.e. guaranteed to be non-null).

File: src/room/GroupCallView.tsx
Comment (daniel-abramov) [reply src/room/GroupCallView.tsx]: (pending on https://github.com/vector-im/element-call/pull/1173#discussion_r1251176236)

File: src/room/GroupCallView.tsx
Comment (daniel-abramov) [reply src/room/GroupCallView.tsx]: (pending on https://github.com/vector-im/element-call/pull/1173#discussion_r1251176236)

File: src/room/GroupCallView.tsx
Comment (daniel-abramov) [reply src/room/GroupCallView.tsx]: (pending on https://github.com/vector-im/element-call/pull/1173#discussion_r1251176236)

File: src/room/InCallView.tsx
Comment (daniel-abramov) [reply src/room/InCallView.tsx]: (pending on https://github.com/vector-im/element-call/pull/1173#discussion_r1251176236)

File: src/room/InCallView.tsx
Comment (daniel-abramov) [reply src/room/InCallView.tsx]: (pending on https://github.com/vector-im/element-call/pull/1173#discussion_r1251176236)

File: src/room/InCallView.tsx
Comment (daniel-abramov) [reply src/room/InCallView.tsx]: (pending on https://github.com/vector-im/element-call/pull/1173#discussion_r1251176236)

File: src/room/GridLayoutMenu.tsx
Comment (robintown) [reply src/room/GridLayoutMenu.tsx]: I don't think so, sadly… but you're very right that it's easy to forget these things. Usually it's not a big deal to leave things unmemoized, but Meta is also experimenting with an auto-memoizing compiler that could improve the developer experience here.

File: src/room/GroupCallLoader.tsx
Comment (robintown) [reply src/room/GroupCallLoader.tsx]: Fantastic :+1:

File: src/room/GroupCallView.tsx
Comment (robintown) [reply src/room/GroupCallView.tsx]: Correct, `widget` can't be modified because it's a `const`.

I think the reason that TypeScript doesn't cast it to a non-null value here is because you're supposed to be able to replace a declaration like `export const widget` with `export let widget` without that being a breaking change in your module's API, and so the type checker has to pretend that `widget` could change to `null` by the time the function is invoked.

Comment (daniel-abramov) [reply src/room/GroupCallView.tsx]: I see, thanks for explaining! Fixed.

**Comment by daniel-abramov** [comment]
**UPD.** So it seems like now all the problems are solved, so the only thing that is left that keeps us from merging (apart from a smoke test that I'll do) is the errors that I mentioned in the description of the issue ([they also fail in the pipeline](https://github.com/vector-im/element-call/actions/runs/5455178543/jobs/9926266319?pr=1173)) - namely the errors that originate from the `matrix-js-sdk`.

The `exclude` pattern in `tsconfig` does not seem to do anything in this case as `matrix-js-sdk` is directly imported and so it looks like it can't be disabled via `exclude` (we have `node_modules` in the `exclude`). The `skipLibCheck: false` option does not seem to do anything particularly useful in this case either.

File: package.json
Comment (dbkr) [inline package.json:13]: Why the rename from lint:js to lint:eslint?

Comment (daniel-abramov) [reply package.json:13]: Oh, I think I did it while comparing `eslint` with `tsc` and thinking that perhaps calling it `lint:eslint` would be slightly more explicit since it only calls `eslint` (`element-web`, `matrix-js-sdk` contain a `lint:js` command that consists of running multiple subcommands, both calling `eslint` though).

But I don't have any strong preference here, so I could revert change if you prefer 🙂

**Review by robintown — APPROVED** [review]
Looks good once the errors are fixed.

For the first one, I see that matrix-js-sdk has the file https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/%40types/another-json.d.ts which is supposed to suppress that error, so I think you can fix this by including that file in our TypeScript config like matrix-react-sdk does: https://github.com/matrix-org/matrix-react-sdk/blob/2a7780052e60f23a8a7a90d549f27dea471e5ccd/tsconfig.json#L22

Not sure about the second error though.

File: package.json
Comment (robintown) [reply package.json:13]: FWIW I like the more explicit name

Comment (dbkr) [reply package.json:13]: Fair enough, seems reasonable, as long as we're consistent between projects.

**Comment by dbkr** [comment]
https://github.com/matrix-org/matrix-rust-sdk/blob/a0f40606d4e379b395b524b3c86d2abc261bf6a8/bindings/matrix-sdk-crypto-js/CHANGELOG.md?plain=1#L34 suggests they've just been added recently, so maybe we need to update js-sdk to pull in a newer copy of the crypto lib.

**Comment by daniel-abramov** [comment]
> Looks good once the errors are fixed.
> 
> For the first one, I see that matrix-js-sdk has the file https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/%40types/another-json.d.ts which is supposed to suppress that error, so I think you can fix this by including that file in our TypeScript config like matrix-react-sdk does: https://github.com/matrix-org/matrix-react-sdk/blob/2a7780052e60f23a8a7a90d549f27dea471e5ccd/tsconfig.json#L22
> 
> Not sure about the second error though.

That did the trick, awesome, thanks!

**Comment by daniel-abramov** [comment]
> https://github.com/matrix-org/matrix-rust-sdk/blob/a0f40606d4e379b395b524b3c86d2abc261bf6a8/bindings/matrix-sdk-crypto-js/CHANGELOG.md?plain=1#L34 suggests they've just been added recently, so maybe we need to update js-sdk to pull in a newer copy of the crypto lib.

You're right! I ran `yarn lint` in the `matrix-js-sdk` and it exhibited the same errors, but once updating the rust crypto SDK in the `package.json`, `yarn lint` succeeds!

Sadly, I can't push to the `livekit` branch, it seems like I lack permissions. Could someone please apply the patch or grant me permissions to push to the branches? 🙂

<details>
The patch is as simple as:

```diff
Author: Daniel Abramov <daniel.abramov@element.io>
Date:   Thu Jul 6 11:15:43 2023 +0100

    Fix linting errors caused by the old crypto SDK

    See https://github.com/vector-im/element-call/pull/1173#issuecomment-1621306979
    for more details!

diff --git a/package.json b/package.json
index 8ffe173ea..470fc65f1 100644
--- a/package.json
+++ b/package.json
@@ -55,7 +55,7 @@
     ],
     "dependencies": {
         "@babel/runtime": "^7.12.5",
-        "@matrix-org/matrix-sdk-crypto-js": "^0.1.0-alpha.10",
+        "@matrix-org/matrix-sdk-crypto-js": "^0.1.0",
         "another-json": "^0.2.0",
         "bs58": "^5.0.0",
         "content-type": "^1.0.4",
diff --git a/yarn.lock b/yarn.lock
index dd15a07b0..1d0622de4 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -1426,10 +1426,10 @@
   dependencies:
     lodash "^4.17.21"

-"@matrix-org/matrix-sdk-crypto-js@^0.1.0-alpha.10":
-  version "0.1.0-alpha.10"
-  resolved "https://registry.yarnpkg.com/@matrix-org/matrix-sdk-crypto-js/-/matrix-sdk-crypto-js-0.1.0-alpha.10.tgz#b6a6395cffd3197ae2e0a88f4eeae8b315571fd2"
-  integrity sha512-8V2NKuzGOFzEZeZVgF2is7gmuopdRbMZ064tzPDE0vN34iX6s3O8A4oxIT7SA3qtymwm3t1yEvTnT+0gfbmh4g==
+"@matrix-org/matrix-sdk-crypto-js@^0.1.0":
+  version "0.1.0"
+  resolved "https://registry.yarnpkg.com/@matrix-org/matrix-sdk-crypto-js/-/matrix-sdk-crypto-js-0.1.0.tgz#766580036d4df12120ded223e13b5640e77db136"
+  integrity sha512-ra/bcFdleC1iRNms2I96UXA0NvQYWpMsHrV5EfJRS7qV1PtnQNvgsvMfjMbkx8QT2ErEmIhsvB5fPCpfp8BSuw==

 "@matrix-org/olm@https://gitlab.matrix.org/api/v4/projects/27/packages/npm/@matrix-org/olm/-/@matrix-org/olm-3.2.14.tgz":
   version "3.2.14"
```

</details>

**Comment by dbkr** [comment]
Can you make a PR to the livekit branch as normal?

**Comment by daniel-abramov** [comment]
> Can you make a PR to the livekit branch as normal?

I tried, but it did not work out. Apparently, I don't have any push permissions.

**Comment by daniel-abramov** [comment]
After [merging `develop` into `livekit` in the Matrix JS SDK](https://github.com/matrix-org/matrix-js-sdk/pull/3573) the original errors disappeared, but now it has a new portion of errors that is similar to the `another-json` errors were resolved following a suggestion from @robintown by including the file from the Matrix JS SDK that contained the `declare module` part.

Now we have the same error but for `uuid`. Should we add it in a similar way to the Matrix JS SDK? (I'm not really sure why it happens at the first place)

cc @toger5

**Comment by dbkr** [comment]
I've submitted https://github.com/matrix-org/matrix-js-sdk/pull/3593 to make these types packages regular dependencies in the js-sdk which I think is the right solution here.

**Pull request merged**
